src/sql/src/plan/statement/with_options.rs

```diff
@@ -97,26 +97,98 @@ macro_rules! with_options {
 
 pub(crate) fn aws_connect_info(
     options: &mut BTreeMap<String, Value>,
-    region: Option<Region>,
-) -> anyhow::Result<AwsConnectInfo> {
-    // todo@jldlaughlin: We should support all (?) variants of AWS authentication.
-    // https://github.com/materializeinc/materialize/issues/1991
+    region: Option<String>,
+) -> anyhow::Result<aws::ConnectInfo> {
     let mut extract = |key| match options.remove(key) {
         Some(Value::String(key)) => Ok(Some(key)),
         Some(_) => bail!("{} must be a string", key),
         _ => Ok(None),
     };
-    let region = match region {
+
+    let region_raw = match region {
         Some(region) => region,
-        None => extract("region")?
-            .map(|r| r.parse())
-            .transpose()?
-            .ok_or_else(|| anyhow::anyhow!("region is required"))?,
+        None => extract("region")?.ok_or_else(|| anyhow!("region is required"))?,
+    };
+
+    let region = match region_raw.parse() {
+        Ok(region) => {
+            // ignore the endpoint option if we're pointing at a
+            // valid, non-custom AWS region
+            let _ = extract("endpoint");
+            region
+        }
+        Err(e) => {
+            // Region's fromstr doesn't support parsing custom regions.
+            // If a Kinesis stream's ARN indicates it exists in a custom
+            // region, support it iff a valid endpoint for the stream
+            // is also provided.
+            match extract("endpoint")? {
+                Some(endpoint) => Region::Custom {
+                    name: region_raw,
+                    endpoint,
+                },
+                _ => bail!(
+                    "Unable to parse AWS region: {}. If providing a custom \
+                         region, an `endpoint` option must also be provided",
+                    e
+                ),
+            }
+        }
     };
-    Ok(AwsConnectInfo {
+
+    aws::ConnectInfo::new(
         region,
-        access_key_id: extract("access_key_id")?,
-        secret_access_key: extract("secret_access_key")?,
-        token: extract("token")?,
-    })
+        extract("access_key_id")?,
+        extract("secret_access_key")?,
+        extract("token")?,
+    )
+}
```

Comment:
Not for now, but this whole function should probably use Matt's new `WITH` option parsing framework.

Reply:
I agree, I ended up here because it wasn't flexible enough when I initially tried. It should be made flexible enough, and possibly some of the refactorings have made it a bit easier.